File: eclipse-scout-core/src/security/access.ts
/*
 * Copyright (c) 2010, 2024 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
import {AccessControl, Permission, PermissionLevel, scout, systems} from '../index';
import $ from 'jquery';

let accessControl: AccessControl = null;

export const access = {

  /**
   * Loads all permissions from the given url to a client cache.
   * @param permissionsUrl The url to fetch the permissions from.
   */
  bootstrap(permissionsUrl: string): JQuery.Promise<any> {
    if (!permissionsUrl) {
      return $.resolvedPromise();
    }
    if (!accessControl) {
      accessControl = scout.create(AccessControl, {permissionsUrl});
    }
    return accessControl.bootstrap();
  },

  /**
   * Loads permissions from the main system using the url configuration of that {@link System}.
   */
  bootstrapSystem(): JQuery.Promise<void> {
    const url = systems.getOrCreate().getEndpointUrl('permissions', 'permissions');
    return access.bootstrap(url);
  },

  tearDown() {
    if (!accessControl) {
      return;
    }
    accessControl.destroy();
    accessControl = null;
  },

  /**
   * Quick check (only local checks) `permission` against granted permissions of the current user.
   */
  quickCheck(permission: string | Permission): boolean {
    return check(permission, true);
  },

  /**
   * Check `permission` against granted permissions of the current user.
   */
  check(permission: string | Permission): JQuery.Promise<boolean> {
    return check(permission);
  },

  /**
   * Returns the granted {@link PermissionLevel} for a given permission instance `permission`.
   * - {@link Permission.Level.UNDEFINED} if `permission` is `null` or in general 'not an {@link Permission}'
   * - {@link Permission.Level.NONE} if no level at all is granted to `permission`
   * - {@link PermissionLevel} if the level can be determined exactly.
   * - {@link Permission.Level.UNDEFINED} if there are multiple granted permission levels possible and there is not enough data in the `permission` contained to determine the result closer.
   */
  getGrantedPermissionLevel(permission: string | Permission): PermissionLevel {
    if (!accessControl) {
      return Permission.Level.UNDEFINED;
    }
    if (!(permission instanceof Permission)) {
      permission = Permission.quick(permission);
    }
    return accessControl.getGrantedPermissionLevel(permission);
  },

  getPermissionsUrl(): string {
    return accessControl && accessControl.permissionsUrl;
  }
};

function check(permission: string | Permission, quick: true): boolean;
function check(permission: string | Permission, quick?: false): JQuery.Promise<boolean>;
function check(permission: string | Permission, quick?: boolean): boolean | JQuery.Promise<boolean> {
  if (!accessControl) {
    return quick ? false : $.resolvedPromise(false);
  }
  if (!(permission instanceof Permission)) {
    permission = Permission.quick(permission);
  }
  return accessControl.check(permission, quick);
}

File: org.eclipse.scout.rt.security/src/main/java/org/eclipse/scout/rt/security/ACCESS.java
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */
package org.eclipse.scout.rt.security;

import java.security.Permission;

import org.eclipse.scout.rt.dataobject.exception.AccessForbiddenException;
import org.eclipse.scout.rt.platform.BEANS;

/**
 * Helper class for checking permissions.
 */
public final class ACCESS {

  private ACCESS() {
  }

  /**
   * Checks permission against granted permissions of current user.
   * <p>
   * <code>
   * if(ACCESS.check(new ReadPersonPermission(personId))){ ... }
   * </code>
   *
   * @param permission
   *     if {@code null}, false is returned
   * @return true if access is granted
   */
  public static boolean check(Permission permission) {
    return BEANS.get(AccessSupport.class).check(permission);
  }

  /**
   * Throws an exception if {@link #check(Permission)} fails for the given permission.
   *
   * @throws AccessForbiddenException
   *     thrown if access is not granted
   * @see IPermission#getAccessCheckFailedMessage()
   */
  public static void checkAndThrow(Permission permission) {
    BEANS.get(AccessSupport.class).checkAndThrow(permission);
  }

  /**
   * @return true if access for any given permission is granted
   */
  public static boolean checkAny(Permission... permissions) {
    return BEANS.get(AccessSupport.class).checkAny(permissions);
  }

  /**
   * Throws an exception if {@link #checkAny(Permission...)} fails for the given permission.
   * <p>
   * The access check failed message is taken from the <em>first</em> permission.
   *
   * @throws AccessForbiddenException
   *     thrown if access is not granted
   * @see IPermission#getAccessCheckFailedMessage()
   */
  public static void checkAnyAndThrow(Permission... permissions) {
    BEANS.get(AccessSupport.class).checkAnyAndThrow(permissions);
  }

  /**
   * @return true if access for all given permission is granted
   */
  public static boolean checkAll(Permission... permissions) {
    return BEANS.get(AccessSupport.class).checkAll(permissions);
  }

  /**
   * Throws an exception if {@link #checkAll(Permission...)} fails for the given permission.
   * <p>
   * The access check failed message is taken from the <em>first</em> permission for which access was not granted.
   *
   * @throws AccessForbiddenException
   *     thrown if access is not granted
   * @see IPermission#getAccessCheckFailedMessage()
   */
  public static void checkAllAndThrow(Permission... permissions) {
    BEANS.get(AccessSupport.class).checkAllAndThrow(permissions);
  }

  /**
   * This methods returns the granted {@link PermissionLevel} for a given permission instance {@code permission}.
   * <ul>
   * <li>{@link PermissionLevel#UNDEFINED} if {@code permission} is {@code null} or in general 'not an
   * {@link IPermission}'
   * <li>{@link PermissionLevel#NONE} if no level at all is granted to {@code permission}
   * <li>{@link PermissionLevel} if the level can be determined exactly.
   * <li>{@link PermissionLevel#UNDEFINED} if there are multiple granted permission levels possible and there is not
   * enough data in the permission contained to determine the result more closer.
   * </ul>
   *
   * @return non null {@link PermissionLevel}
   * @see IPermissionCollection#getGrantedPermissionLevel(IPermission)
   */
  public static PermissionLevel getGrantedPermissionLevel(IPermission permission) {
    return BEANS.get(AccessSupport.class).getGrantedPermissionLevel(permission);
  }
}
